File: Netsuflow/openfx/src/DiagnosticFrame.cpp
#include "DiagnosticFrame.hpp"

#include <limits>

namespace netsuflow {
namespace {

// Frame is masked to 24 bits inside the pattern so the same arithmetic is exact
// in IEEE-754 doubles, which is what the JavaScript mirror uses.
constexpr std::uint32_t kPatternFrameMask = 0x00FFFFFFu;

constexpr std::uint32_t kCounterCells = 16;

inline void patternPixel(std::uint32_t x, std::uint32_t y, std::uint32_t frame,
                         std::uint8_t* out) noexcept {
  const std::uint32_t f = frame & kPatternFrameMask;
  out[0] = static_cast<std::uint8_t>((x * 7u + f * 13u) & 0xFFu);
  out[1] = static_cast<std::uint8_t>((y * 11u + f * 29u) & 0xFFu);
  out[2] = static_cast<std::uint8_t>(((x ^ y) + f * 3u) & 0xFFu);
  out[3] = 255u;
}

inline std::uint32_t counterBandHeight(std::uint32_t height) noexcept {
  const std::uint32_t tenth = height / 10u;
  const std::uint32_t band = tenth > 8u ? tenth : 8u;
  return band < height ? band : height;
}

}  // namespace

bool isValidFrameSpec(const FrameSpec& spec) noexcept {
  return spec.width > 0 && spec.height > 0 && spec.width <= kMaxFrameDimension &&
         spec.height <= kMaxFrameDimension;
}

std::size_t diagnosticFrameByteSize(const FrameSpec& spec) noexcept {
  if (!isValidFrameSpec(spec)) return 0;
  const std::uint64_t bytes = static_cast<std::uint64_t>(spec.width) *
                              static_cast<std::uint64_t>(spec.height) *
                              static_cast<std::uint64_t>(kBytesPerPixelRgba8);
  if (bytes > static_cast<std::uint64_t>((std::numeric_limits<std::size_t>::max)())) return 0;
  return static_cast<std::size_t>(bytes);
}

bool writeDiagnosticFrame(const FrameSpec& spec, std::uint8_t* dest,
                          std::size_t destSize) noexcept {
  const std::size_t needed = diagnosticFrameByteSize(spec);
  if (needed == 0 || dest == nullptr || destSize < needed) return false;

  for (std::uint32_t y = 0; y < spec.height; ++y) {
    std::uint8_t* row = dest + static_cast<std::size_t>(y) * spec.width * kBytesPerPixelRgba8;
    for (std::uint32_t x = 0; x < spec.width; ++x) {
      patternPixel(x, y, spec.frame, row + static_cast<std::size_t>(x) * kBytesPerPixelRgba8);
    }
  }

  // Visible binary frame counter across the top band: cell i is lit when bit i
  // of the low 16 bits of the frame number is set. Readable by eye in Fusion,
  // which is what T01 needs from a generator that cannot rasterize text.
  const std::uint32_t band = counterBandHeight(spec.height);
  const std::uint32_t cellWidth = spec.width / kCounterCells;
  if (cellWidth > 0) {
    for (std::uint32_t cell = 0; cell < kCounterCells; ++cell) {
      const bool lit = ((spec.frame >> cell) & 1u) != 0u;
      const std::uint8_t value = lit ? 255u : 16u;
      const std::uint32_t x0 = cell * cellWidth;
      const std::uint32_t x1 = (cell + 1 == kCounterCells) ? spec.width : x0 + cellWidth;
      for (std::uint32_t y = 0; y < band; ++y) {
        std::uint8_t* row = dest + static_cast<std::size_t>(y) * spec.width * kBytesPerPixelRgba8;
        for (std::uint32_t x = x0; x < x1; ++x) {
          std::uint8_t* px = row + static_cast<std::size_t>(x) * kBytesPerPixelRgba8;
          px[0] = value;
          px[1] = value;
          px[2] = value;
          px[3] = 255u;
        }
      }
    }
  }

  // Machine-readable frame marker: big-endian frame number in the red channel of
  // the first four pixels of row 0.
  if (spec.width >= 4) {
    for (std::uint32_t i = 0; i < 4; ++i) {
      const std::uint32_t shift = (3u - i) * 8u;
      dest[static_cast<std::size_t>(i) * kBytesPerPixelRgba8] =
          static_cast<std::uint8_t>((spec.frame >> shift) & 0xFFu);
    }
  }

  return true;
}

std::vector<std::uint8_t> makeDiagnosticFrame(const FrameSpec& spec) {
  const std::size_t needed = diagnosticFrameByteSize(spec);
  if (needed == 0) return {};
  std::vector<std::uint8_t> pixels(needed);
  if (!writeDiagnosticFrame(spec, pixels.data(), pixels.size())) return {};
  return pixels;
}

std::uint32_t frameMarker(const std::uint8_t* pixels, std::size_t size,
                          const FrameSpec& spec) noexcept {
  if (pixels == nullptr || spec.width < 4) return kNoFrameMarker;
  const std::size_t needed = diagnosticFrameByteSize(spec);
  if (needed == 0 || size < needed) return kNoFrameMarker;
  std::uint32_t value = 0;
  for (std::uint32_t i = 0; i < 4; ++i) {
    value = (value << 8) | pixels[static_cast<std::size_t>(i) * kBytesPerPixelRgba8];
  }
  return value;
}

std::uint32_t frameMarker(const std::vector<std::uint8_t>& pixels,
                          const FrameSpec& spec) noexcept {
  return frameMarker(pixels.data(), pixels.size(), spec);
}

}  // namespace netsuflow

File: Netsuflow/prototypes/fake-renderer/diagnosticFrame.mjs
// JavaScript mirror of openfx/src/DiagnosticFrame.cpp.
//
// Both implementations must produce byte-identical output for the same spec.
// That equality is what proves, in T03, that pixels crossed the bridge unchanged
// without Remotion being involved at all. `diagnosticFrame.test.mjs` pins the
// arithmetic; the end-to-end test compares the two implementations directly.

export const MAX_FRAME_DIMENSION = 16384;
export const BYTES_PER_PIXEL = 4;
export const NO_FRAME_MARKER = 0xffffffff;

const PATTERN_FRAME_MASK = 0x00ffffff;
const COUNTER_CELLS = 16;

export function isValidFrameSpec({ width, height }) {
  return (
    Number.isInteger(width) &&
    Number.isInteger(height) &&
    width > 0 &&
    height > 0 &&
    width <= MAX_FRAME_DIMENSION &&
    height <= MAX_FRAME_DIMENSION
  );
}

export function diagnosticFrameByteSize(spec) {
  if (!isValidFrameSpec(spec)) return 0;
  return spec.width * spec.height * BYTES_PER_PIXEL;
}

function counterBandHeight(height) {
  const tenth = Math.floor(height / 10);
  const band = tenth > 8 ? tenth : 8;
  return band < height ? band : height;
}

export function makeDiagnosticFrame({ width, height, frame }) {
  const spec = { width, height, frame };
  const size = diagnosticFrameByteSize(spec);
  if (size === 0) return null;

  const pixels = Buffer.allocUnsafe(size);
  const f = frame & PATTERN_FRAME_MASK;

  for (let y = 0; y < height; y += 1) {
    const rowStart = y * width * BYTES_PER_PIXEL;
    for (let x = 0; x < width; x += 1) {
      const offset = rowStart + x * BYTES_PER_PIXEL;
      pixels[offset] = (x * 7 + f * 13) & 0xff;
      pixels[offset + 1] = (y * 11 + f * 29) & 0xff;
      pixels[offset + 2] = ((x ^ y) + f * 3) & 0xff;
      pixels[offset + 3] = 255;
    }
  }

  // Visible binary frame counter across the top band.
  const band = counterBandHeight(height);
  const cellWidth = Math.floor(width / COUNTER_CELLS);
  if (cellWidth > 0) {
    for (let cell = 0; cell < COUNTER_CELLS; cell += 1) {
      const lit = ((frame >>> cell) & 1) !== 0;
      const value = lit ? 255 : 16;
      const x0 = cell * cellWidth;
      const x1 = cell + 1 === COUNTER_CELLS ? width : x0 + cellWidth;
      for (let y = 0; y < band; y += 1) {
        const rowStart = y * width * BYTES_PER_PIXEL;
        for (let x = x0; x < x1; x += 1) {
          const offset = rowStart + x * BYTES_PER_PIXEL;
          pixels[offset] = value;
          pixels[offset + 1] = value;
          pixels[offset + 2] = value;
          pixels[offset + 3] = 255;
        }
      }
    }
  }

  // Machine-readable frame marker in the red channel of the first four pixels.
  if (width >= 4) {
    for (let i = 0; i < 4; i += 1) {
      const shift = (3 - i) * 8;
      pixels[i * BYTES_PER_PIXEL] = (frame >>> shift) & 0xff;
    }
  }

  return pixels;
}

export function frameMarker(pixels, spec) {
  if (!pixels || spec.width < 4) return NO_FRAME_MARKER;
  const size = diagnosticFrameByteSize(spec);
  if (size === 0 || pixels.length < size) return NO_FRAME_MARKER;
  let value = 0;
  for (let i = 0; i < 4; i += 1) {
    value = ((value << 8) | pixels[i * BYTES_PER_PIXEL]) >>> 0;
  }
  return value;
}
